### PR TITLE
Fix for missing k8s logs

### PR DESCRIPTION
**Pull request by jzh0u**
Fix for the missing pod logs in the kubernetes driver.

We didn't check to make sure io.Copy() was writing anything, so sometimes it would finish before the pod actually wrote anything. Added a few checks to remedy this.

To reproduce bug:
Run `go test -tags=integration -count=20 ./driver/kubernetes`
See below fail at least once
image

### CODE REVIEW

File: driver/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -316,8 +316,15 @@ func (k *Driver) streamPodLogs(options metav1.ListOptions, out io.Writer, done c
...
-			io.Copy(out, reader)
+			n, err := io.Copy(out, reader)
 			reader.Close()
+			if err != nil {
+				continue
+			}
+			if n == 0 {
+				streamedLogs[podName] = false
+				continue
+			}
```
Comment (jlegrone) [inline driver/kubernetes/kubernetes.go]: This may cause an infinite loop if an invocation image does not have any output. Can we retry up to a certain number of times to get logs, and then exit?

File: driver/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -316,8 +316,15 @@ func (k *Driver) streamPodLogs(options metav1.ListOptions, out io.Writer, done c
 			// We successfully connected to the pod, so mark it as having streamed logs.
 			streamedLogs[podName] = true
 			// Block the loop until all logs from the pod have been processed.
-			io.Copy(out, reader)
+			n, err := io.Copy(out, reader)
 			reader.Close()
+			if err != nil {
+				continue
```
Comment (glyn) [inline driver/kubernetes/kubernetes.go]: If there's a (transient) error, it is possible that no logs were copied and yet we continue with the pod marked as having streamed logs, so a retry won't occur.

File: driver/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -316,8 +316,15 @@ func (k *Driver) streamPodLogs(options metav1.ListOptions, out io.Writer, done c
 			// We successfully connected to the pod, so mark it as having streamed logs.
 			streamedLogs[podName] = true
```
Comment (glyn) [inline driver/kubernetes/kubernetes.go]: Wouldn't it be better to set this to true if and only if (all?) the logs have been streamed?

File: driver/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -346,6 +368,7 @@ func generateLabels(op *driver.Operation) map[string]string {
 		"cnab.io/installation": op.Installation,
 		"cnab.io/action":       op.Action,
 		"cnab.io/revision":     op.Revision,
+		"random_string":        strconv.Itoa(rand.Intn(1000)),
```
Comment (glyn) [inline driver/kubernetes/kubernetes.go]: How is this relevant to this PR?

File: driver/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -291,34 +293,54 @@ func (k *Driver) streamPodLogs(options metav1.ListOptions, out io.Writer, done c
...
-				Follow:    true,
-			})
-			reader, err := req.Stream()
-			// There was an error connecting to the pod, so continue the loop and attempt streaming
-			// logs again next time there is an event for the same pod.
-			if err != nil {
-				continue
+
+			timeout := 0
+			for (streamedLogs[podName] < numBackoffLoops) && (timeout < numBackoffLoops) {
```
Comment (glyn) [inline driver/kubernetes/kubernetes.go]: Unfortunately, the additional logic is becoming more convoluted rather than simpler.

File: driver/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -291,34 +293,54 @@ func (k *Driver) streamPodLogs(options metav1.ListOptions, out io.Writer, done c
...
+				req := k.pods.GetLogs(podName, &v1.PodLogOptions{
+					Container: k8sContainerName,
+					Follow:    true,
+				})
+				reader, err := req.Stream()
+				// There was an error connecting to the pod, so continue the loop and attempt streaming
+				// logs again next time there is an event for the same pod.
+				if err != nil {
+					timeout = timeout + 1
```
Comment (glyn) [inline driver/kubernetes/kubernetes.go]: There isn't a delay before trying again, so a transient problem might not have chance to clear.

File: driver/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -291,34 +293,54 @@ func (k *Driver) streamPodLogs(options metav1.ListOptions, out io.Writer, done c
...
+				// Block the loop until all logs from the pod have been processed.
+				n, err := io.Copy(out, reader)
+				reader.Close()
+				if err != nil {
+					continue
+				}
+				// There is a chance where we have connected to the pod, but it has yet to write something.
+				// In that case, we continue to to keep streaming until it does.
+				if n == 0 {
+					continue
```
Comment (glyn) [inline driver/kubernetes/kubernetes.go]: Here we _know_ we haven't copied any logs, but `streamedLogs[podName]` has still been incremented, which seems incorrect.

**Comment by glyn** [comment]
@jzh0u Thanks for trying again, but the logic is becoming pretty convoluted. It might be best to try to strip it down to the absolute minimum required to fix the problem.

Also, regardless of this PR, you've found a valid bug, so it would probably be helpful to submit an issue.

File: driver/kubernetes/kubernetes.go (hunk 4 above)
Comment (jzh0u) [reply driver/kubernetes/kubernetes.go]: It's a temporary fix for a pod race condition I was running into. Sometimes, I ended up streaming previous pods that haven't finished terminating because the labels are the same between trials. I'm currently looking into a better way of doing this, but just wanted to quickly fix the issue.

File: driver/kubernetes/kubernetes.go (hunk 7 above)
Comment (jzh0u) [reply driver/kubernetes/kubernetes.go]: But in the case that there are no logs, if we don't increment `streamedLogs[podName]` the whole thing will hang

**Comment by glyn** [comment]
@vdice Please could you advise what's gone wrong with the `tests` check? I can't find any way to debug this.
